src/zos-mkdtemp.c

```diff
@@ -0,0 +1,105 @@
+//////////////////////////////////////////////////////////////////////////////
+//// Licensed Materials - Property of IBM
+//// ZOSLIB
+//// (C) Copyright IBM Corp. 2020. All Rights Reserved.
+//// US Government Users Restricted Rights - Use, duplication
+//// or disclosure restricted by GSA ADP Schedule Contract with IBM Corp.
+/////////////////////////////////////////////////////////////////////////////////
+
+#include <stdio.h>
+#include <string.h>
+#include <time.h>
+#include <stdlib.h>
+#include <errno.h>
+#include <sys/stat.h>
+
+typedef enum {
+  cp_ebcdic,
+  cp_ascii
+} CodePage;
+
+#define XXXXXX "XXXXXX"
+#define ValidChars "0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ_.,-"
+typedef struct CodePage {
+  char xes[sizeof(XXXXXX)];
+  char validChars[sizeof(ValidChars)];
+}  CodePage_t;
+
+#pragma convert("IBM-1047")
+static const CodePage_t CharTraitsEbcdic = {
+   XXXXXX,
+   ValidChars
+};
+#pragma convert(pop)
+
+#pragma convert("ISO8859-1")
+static const CodePage_t CharTraitsAscii = {
+   XXXXXX,
+   ValidChars
+};
+#pragma convert(pop)
+
+enum {
+  TempLen = sizeof(XXXXXX)-1
+};
+
+char *real_mkdtemp(CodePage_t cp, char *templ) {
+  if (!templ) {
+    errno=EINVAL;
+    return NULL;
+  }
+  int len = strlen(templ);
+
+  if (len <TempLen || strncmp(templ+len-TempLen,cp.xes ,TempLen)!=0) {
+    errno=EINVAL;
+    return NULL;
+  }
+#define validChars cp.validChars
+  srand(time(NULL));
+#define RANDOM_INT(max) (1 + rand()/((RAND_MAX + 1u) / (max) ))
+#define UPDATE_CHAR(n) templ[len-n+n-1] = validChars[(s##n+c##n)%(sizeof validChars)]
```

Comment:
minor: I wonder if an array (s and c) and using n as the index to s and c would be cleaner than a macro.

Reply:
I think I would still define UPDATE_CHAR since we need to preform the expression multiple times.  Using an array for s & c would work.  I'm not sure it would make the better.  The macro would be easier to read but the other code would change from `s1` to `s[0]` and everything would have to become zero based.

I did notice a bug though `len-n+n-1` isn't right.  It should be `len-TempLen+n-1`.  I'm not sure how that got in there but I'll fix.